geps/gep-2162.md

```diff
@@ -0,0 +1,108 @@
+
+## Goals
+
+* Improve UX by enabling users to easily see what features the implementation (GatewayClass) support.
+
+* Automatically run conformance tests based on the supported features populated in GatewayClass status.
+
+* Provide foundation for tools to block or warn when unsupported features are used.
+
+## Non-Goals
+
+* Validate correctness of supported features published by the implementation. Meaning we don't intend to verify whether the supported features reported by the implementation are indeed supported.
+
+## Introduction
+
+The current [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185 is only used to store conditions the controller publishes.
+
+Partnered with the [Conformance Profiles](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-1709.md) work, we want to:
+
+1. Improve UX by enabling users to easily see what features the implementation(GatewayClass) support.
+1. Automatically run conformance tests based on the supported features populated in GatewayClass status.
+1. Potentially build tooling to block or warn when unsupported features are used (more under [Future Work](#future-work)).
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **must** publish the supported features before Accepting the GatewayClass.
+
+Implementations are free to decide how they manage this information - common approaches could be either to maintain static lists of supported features OR generate them from their conformance tests results/reports as part of their CI.
+
+## API
+
+This GEP proposes API changes describes as follow:
+
+* Update the `GatewayClassStatus` struct to include a string-represented list of `SupportedFeatures`.
+
+
+```go
+// GatewayClassStatus is the current status for the GatewayClass.
+type GatewayClassStatus struct {
+    // Conditions is the current status from the controller for
+    // this GatewayClass.
+    //
+    // Controllers should prefer to publish conditions using values
+    // of GatewayClassConditionType for the type of each Condition.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    // +kubebuilder:validation:MaxItems=8
+    // +kubebuilder:default={{type: "Accepted", status: "Unknown", message: "Waiting for controller", reason: "Pending", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+    // SupportedFeatures is the features the GatewayClass support.
+    // <gateway:experimental>
+    // +kubebuilder:validation:MaxItems=64
+    SupportedFeatures []string `json:"SupportedFeatures,omitempty"`
```

Comment:
In a zoom, @LiorLieberman @robscott and myself discussed the possibility that this type would be something _other_ than a `[]string`. We want this list to be human readable in addition to machine readable, and so at minimum I think since features will be in the `<*Route><Feature>` format, it could be very helpful to human readers if this was sorted in lexicographical order.

```suggestion
    // SupportedFeatures is the features the GatewayClass support.
    //
    // Implements MUST sort this list in lexicographical order to help make it easier
    // to parse for human readers.
    //
    // <gateway:experimental>
    // +kubebuilder:validation:MaxItems=64
    SupportedFeatures []string `json:"SupportedFeatures,omitempty"`
```

Reply:
Having the features split by type probably does make sense, I think we might end up having some top-level features as well though ("Allow cross-namespace references using ReferenceGrant" feels like a likely candidate).

Reply:
Wouldn't we just bucket that under `ReferenceGrant`? Some implementations today (I know of 2) explicitly do not support `ReferenceGrant` at all.

Reply:
`ReferenceGrant` is a good point, _technically_ it's like a feature of every route type, but we wouldn't necessarily want to list it that way :thinking: would we?

Reply:
Resolving this in favour of Robs PR here https://github.com/kubernetes-sigs/gateway-api/pull/2323 and the new formatting section I added. 

Feel free to reopen if you feel we need to discuss this more.